File: packages/contracts/src/envelope.ts
import { z } from "zod/v4";
import { Id, SchemaVersion, Timestamp, SCHEMA_VERSION } from "./common";

/**
 * Transport envelope wrapping every event published to the NATS backbone
 * (P05-T002). Producers never publish a bare payload — they wrap it so that
 * consumers, the replay utility, and the inspection CLI all see the same
 * metadata regardless of which stream/subject the event arrived on.
 *
 * Mirrored in Rust at `crates/event_model/src/envelope.rs`.
 *
 * - `event_id`    — unique id for this event (dedup, idempotency).
 * - `schema_version` — wire-format version of the envelope (`SCHEMA_VERSION`).
 * - `trace_id`    — correlates events across the pipeline; defaults to `event_id`
 *                   at the origin and is propagated by downstream producers.
 * - `source`      — logical producer id, e.g. `ingestion`, `features`.
 * - `emitted_at`  — when the producer emitted the envelope (server clock).
 * - `payload_type`— canonical name of the payload contract (see {@link PAYLOAD_TYPES}).
 * - `payload`     — the contract instance; typed via {@link envelopeOf}.
 */
export const EventEnvelope = z.object({
  event_id: Id,
  schema_version: SchemaVersion,
  trace_id: Id,
  source: Id,
  emitted_at: Timestamp,
  payload_type: z.string().min(1),
  payload: z.unknown(),
});
export type EventEnvelope = z.infer<typeof EventEnvelope>;

/**
 * Canonical payload-type names. Open by design (DLQ and future events add
 * their own), but these are the names the stream topology expects.
 */
export const PAYLOAD_TYPES = {
  RawMarketEvent: "RawMarketEvent",
  NormalizedMarketEvent: "NormalizedMarketEvent",
  FeatureSnapshot: "FeatureSnapshot",
  AnomalyEvent: "AnomalyEvent",
  ContextPacket: "ContextPacket",
  Briefing: "Briefing",
  Decision: "Decision",
  SystemHealth: "SystemHealth",
} as const;
export type PayloadType = (typeof PAYLOAD_TYPES)[keyof typeof PAYLOAD_TYPES];

/**
 * Build an envelope schema with a typed `payload`. Use this to validate that an
 * envelope both has the right metadata and carries the expected contract:
 *
 * ```ts
 * const RawEnvelope = envelopeOf(RawMarketEvent);
 * RawEnvelope.parse(incoming); // payload is RawMarketEvent
 * ```
 */
export function envelopeOf<T extends z.ZodTypeAny>(payload: T) {
  return EventEnvelope.extend({ payload });
}

/** Fields a caller supplies to {@link makeEnvelope}; the rest are filled in. */
export interface MakeEnvelopeInput<T> {
  source: string;
  payload_type: string;
  payload: T;
  /** Defaults to a fresh `event_id` (origin of a new trace). */
  trace_id?: string;
  /** Defaults to a generated UUID. Inject for deterministic fixtures/tests. */
  event_id?: string;
  /** Defaults to the current time. Inject for deterministic fixtures/tests. */
  emitted_at?: string;
}

/**
 * Construct a well-formed envelope, filling `event_id`, `emitted_at`,
 * `schema_version`, and defaulting `trace_id` to `event_id` when omitted.
 * Ids/timestamps are injectable so fixtures and tests stay deterministic.
 */
export function makeEnvelope<T>(input: MakeEnvelopeInput<T>): EventEnvelope & { payload: T } {
  const event_id = input.event_id ?? crypto.randomUUID();
  return {
    event_id,
    schema_version: SCHEMA_VERSION,
    trace_id: input.trace_id ?? event_id,
    source: input.source,
    emitted_at: input.emitted_at ?? new Date().toISOString(),
    payload_type: input.payload_type,
    payload: input.payload,
  };
}

File: crates/event_model/src/envelope.rs
//! Transport envelope for the NATS event backbone (P05-T002).
//!
//! Rust mirror of `packages/contracts/src/envelope.ts`. Producers wrap every
//! payload in an [`Envelope`] so consumers, the replay utility, and the
//! inspection CLI see consistent metadata regardless of stream/subject.

use serde::{Deserialize, Serialize};
use time::format_description::well_known::Rfc3339;
use time::OffsetDateTime;
use uuid::Uuid;

/// Wire-format version of the envelope. Mirrors `SCHEMA_VERSION` in the TS
/// contracts (`packages/contracts/src/common.ts`).
pub const SCHEMA_VERSION: u32 = 1;

/// A transport envelope carrying a typed `payload` plus pipeline metadata.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Envelope<T> {
    /// Unique id for this event (dedup, idempotency).
    pub event_id: String,
    /// Wire-format version of the envelope.
    pub schema_version: u32,
    /// Correlates events across the pipeline; defaults to `event_id` at origin.
    pub trace_id: String,
    /// Logical producer id, e.g. `ingestion`, `features`.
    pub source: String,
    /// RFC-3339 timestamp of when the producer emitted the envelope.
    pub emitted_at: String,
    /// Canonical name of the payload contract, e.g. `RawMarketEvent`.
    pub payload_type: String,
    /// The contract instance.
    pub payload: T,
}

impl<T> Envelope<T> {
    /// Build an envelope, generating a fresh `event_id` (UUID v4) and stamping
    /// `emitted_at` with the current UTC time. `trace_id` defaults to the
    /// generated `event_id` (the origin of a new trace).
    #[must_use]
    pub fn new(source: impl Into<String>, payload_type: impl Into<String>, payload: T) -> Self {
        let event_id = Uuid::new_v4().to_string();
        Self {
            trace_id: event_id.clone(),
            event_id,
            schema_version: SCHEMA_VERSION,
            source: source.into(),
            emitted_at: now_rfc3339(),
            payload_type: payload_type.into(),
            payload,
        }
    }

    /// Set the trace id (propagate an existing trace through this producer).
    #[must_use]
    pub fn with_trace_id(mut self, trace_id: impl Into<String>) -> Self {
        self.trace_id = trace_id.into();
        self
    }

    /// Fully-explicit constructor for deterministic fixtures/tests.
    #[must_use]
    pub fn with(
        event_id: impl Into<String>,
        trace_id: impl Into<String>,
        source: impl Into<String>,
        emitted_at: impl Into<String>,
        payload_type: impl Into<String>,
        payload: T,
    ) -> Self {
        Self {
            event_id: event_id.into(),
            schema_version: SCHEMA_VERSION,
            trace_id: trace_id.into(),
            source: source.into(),
            emitted_at: emitted_at.into(),
            payload_type: payload_type.into(),
            payload,
        }
    }
}

/// Current UTC time as an RFC-3339 string. Falls back to the Unix epoch if the
/// platform clock is unavailable or formatting fails (never panics — the crate
/// denies `unwrap`/`expect`).
fn now_rfc3339() -> String {
    OffsetDateTime::now_utc()
        .format(&Rfc3339)
        .unwrap_or_else(|_| "1970-01-01T00:00:00Z".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::{Deserialize, Serialize};

    #[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
    struct Ping {
        msg: String,
    }

    #[test]
    fn new_fills_metadata_and_defaults_trace_id() {
        let env = Envelope::new("ingestion", "Ping", Ping { msg: "hi".into() });
        assert_eq!(env.schema_version, SCHEMA_VERSION);
        assert!(!env.event_id.is_empty());
        assert_eq!(env.trace_id, env.event_id);
        assert_eq!(env.source, "ingestion");
        assert_eq!(env.payload_type, "Ping");
    }

    #[test]
    fn roundtrips_through_json() {
        let env = Envelope::with(
            "evt-1",
            "trace-1",
            "ingestion",
            "2026-06-08T00:00:00Z",
            "Ping",
            Ping { msg: "hi".into() },
        );
        let bytes = serde_json::to_vec(&env).expect("serialize");
        let back: Envelope<Ping> = serde_json::from_slice(&bytes).expect("deserialize");
        assert_eq!(env, back);
    }

    #[test]
    fn with_trace_id_overrides() {
        let env =
            Envelope::new("features", "Ping", Ping { msg: "x".into() }).with_trace_id("trace-99");
        assert_eq!(env.trace_id, "trace-99");
        assert_ne!(env.event_id, "trace-99");
    }
}
